src/bin/cli.rs

```diff
@@ -0,0 +1,637 @@
+                     version.minor,
+                     version.patch);
+            if let Some(ref cfg) = version.cfg_info {
+                if let Some(ref ci) = cfg.commit_info {
+                    println!("commit-hash: {}", ci.commit_hash);
+                    println!("commit-date: {}", ci.commit_date);
+                }
+            }
+        }
+        return Ok(());
+    }
+
+    if let Some(ref code) = args.value_of("explain") {
+        let mut procss = config.rustc()?.process();
+        procss.arg("--explain").arg(code).exec()?;
+        return Ok(());
+    }
+
+    if args.is_present("list") {
+        println!("Installed Commands:");
+        for command in list_commands(config) {
+            let (command, path) = command;
+            if is_verbose {
+                match path {
+                    Some(p) => println!("    {:<20} {}", command, p),
+                    None => println!("    {:<20}", command),
+                }
+            } else {
+                println!("    {}", command);
+            }
+        }
+        return Ok(());
+    }
+
+    if args.subcommand_name().is_none() {
+        cli().print_help()?;
+        return Ok(());
+    }
+
+    execute_subcommand(config, args)
+}
+
+fn execute_subcommand(config: &mut Config, args: ArgMatches) -> CliResult {
+    config.configure(
+        args.occurrences_of("verbose") as u32,
+        if args.is_present("quite") { Some(true) } else { None },
+        &args.value_of("color").map(|s| s.to_string()),
+        args.is_present("frozen"),
+        args.is_present("locked"),
+        &args.values_of_lossy("unstable-features").unwrap_or_default(),
+    )?;
+
+    match args.subcommand() {
+        ("bench", Some(args)) => {
+            let ws = args.workspace(config)?;
+            let mut compile_opts = args.compile_options(config, CompileMode::Bench)?;
+            compile_opts.release = true;
+
+            let ops = ops::TestOptions {
+                no_run: args.is_present("no-run"),
+                no_fail_fast: args.is_present("no-fail-fast"),
+                only_doc: false,
+                compile_opts,
+            };
+
+            let mut bench_args = vec![];
+            bench_args.extend(args.value_of("BENCHNAME").into_iter().map(|s| s.to_string()));
+            bench_args.extend(args.values_of("args").unwrap_or_default().map(|s| s.to_string()));
+
+            let err = ops::run_benches(&ws, &ops, &bench_args)?;
+            match err {
+                None => Ok(()),
+                Some(err) => {
+                    Err(match err.exit.as_ref().and_then(|e| e.code()) {
+                        Some(i) => CliError::new(format_err!("bench failed"), i),
+                        None => CliError::new(err.into(), 101)
+                    })
+                }
+            }
+        }
+        ("build", Some(args)) => {
+            let mut ws = args.workspace(config)?;
+            if config.cli_unstable().avoid_dev_deps {
```

Comment:
Ah this was only present on `build` earlier but I think we'll probably want to fold it into the `workspace` helper method perhaps?

Reply:
Done! Sort of the reason why I want to get rid of code duplication :)